internal/providers/nscloud/api/rpc.go

```diff
@@ -292,45 +293,58 @@ func ListClusters(ctx context.Context, api API) (*KubernetesClusterList, error)
 
 func ExchangeToken(ctx context.Context, scopes ...string) (string, error) {
 	return tasks.Return(ctx, tasks.Action("nscloud.exchange-token"), func(ctx context.Context) (string, error) {
-		// Check if there is already tenant token stored.
-		tenantToken, err := auth.LoadTenantToken(ctx)
+		var username string
+		userAuth, err := auth.LoadUser()
 		if err == nil {
-			// If no scopes provided we can immediately return the token.
-			if len(scopes) == 0 {
-				return tenantToken.TenantToken, nil
-			}
+			username = userAuth.Username
+		} else if !(errors.Is(err, auth.ErrRelogin) && os.Getenv("GITHUB_ACTIONS") == "true") {
+			// XXX: when running in Github Actions no user set otherwise return the error.
+			return "", err
+		}
+
+		scopedToken, err := auth.LoadTenantToken(ctx, username, scopes)
+		if err == nil {
+			// The tenant token with requested scopes is in cache.
+			return scopedToken.TenantToken, nil
+		}
 
+		var token string
+		// If no scoped token cached, then first try exchange unscoped tenant token and otherwise use user token.
+		tenantToken, err := auth.LoadTenantToken(ctx, username, nil)
```

Comment:
Username is a past concept, lets not carry it over. :-)

Reply:
We discussed the use-case when user logins as a different user, I still need some identity to attach tenant tokens to.